src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HostingDiagnosticListener.cs

```diff
@@ -34,31 +46,99 @@ public HostingDiagnosticListener(TelemetryClient client, ICorrelationIdLookupHel
+                }
+
+                var currentActivity = Activity.Current;
+                var isActivityCreatedFromRequestIdHeader = false;
+
+                StringValues xmsRequestRootId;
+                if (currentActivity.ParentId != null)
+                {
+                    isActivityCreatedFromRequestIdHeader = true;
+                }
+                else if (httpContext.Request.Headers.TryGetValue(RequestResponseHeaders.StandardRootIdHeader, out xmsRequestRootId))
+                {
+                    var activity = new Activity(ActivityCreatedByHostingDiagnosticListener);
+                    activity.SetParentId(xmsRequestRootId);
+                    activity.Start();
+                    httpContext.Features.Set(activity);
+                }
+
+                var requestTelemetry = InitializeRequestTelemetryFromActivity(httpContext, Activity.Current, isActivityCreatedFromRequestIdHeader, Stopwatch.GetTimestamp());
+                SetAppIdInResponseHeader(httpContext, requestTelemetry);
+            }
+        }
+
+        /// <summary>
+        /// Diagnostic event handler method for 'Microsoft.AspNetCore.Hosting.HttpRequestIn.Stop' event.
+        /// </summary>
+        [DiagnosticName("Microsoft.AspNetCore.Hosting.HttpRequestIn.Stop")]
+        public void OnHttpRequestInStop(HttpContext httpContext)
+        {
+            EndRequest(httpContext, Stopwatch.GetTimestamp());
+        }
+
+        /// <summary>
         /// Diagnostic event handler method for 'Microsoft.AspNetCore.Hosting.BeginRequest' event.
         /// </summary>
         [DiagnosticName("Microsoft.AspNetCore.Hosting.BeginRequest")]
         public void OnBeginRequest(HttpContext httpContext, long timestamp)
         {
-            if (this.client.IsEnabled())
+            if (this.client.IsEnabled() && !IsAspNetCore20)
             {
-                var requestTelemetry = new RequestTelemetry
-                {
-                    Id = httpContext.TraceIdentifier
-                };
-                this.client.Initialize(requestTelemetry);
-                requestTelemetry.Start(timestamp);
-                httpContext.Features.Set(requestTelemetry);
-                IHeaderDictionary responseHeaders = httpContext.Response?.Headers;
-                if (responseHeaders != null &&
-                    !string.IsNullOrEmpty(requestTelemetry.Context.InstrumentationKey) &&
-                    (!responseHeaders.ContainsKey(RequestResponseHeaders.RequestContextHeader) || HttpHeadersUtilities.ContainsRequestContextKeyValue(responseHeaders, RequestResponseHeaders.RequestContextTargetKey)))
+                var activity = new Activity(ActivityCreatedByHostingDiagnosticListener);
+                var isActivityCreatedFromRequestIdHeader = false;
+
+                StringValues requestId;
+                StringValues standardRootId;
+                if (httpContext.Request.Headers.TryGetValue(RequestResponseHeaders.RequestIdHeader, out requestId))
```

Comment:
The instrumentation code will always send the httpContext, but I'm not sure whether it could be null at there. Since there is never a null check for httpContext, I think it's almost guaranteed.